cuda/base/kernel_launch_reduction.cuh

```diff
@@ -0,0 +1,535 @@
+                             ? warp_partial[threadIdx.x]
+                             : init,
+                         op);
+        if (threadIdx.x == 0) {
+            storage[blockIdx.x] = finalize(partial);
+        }
+    }
+}
+
+
+template <typename ValueType, typename KernelFunction, typename ReductionOp,
+          typename FinalizeOp, typename... KernelArgs>
+__global__ __launch_bounds__(
+    default_block_size) void generic_kernel_reduction_2d(int64 rows, int64 cols,
+                                                         KernelFunction fn,
+                                                         ReductionOp op,
+                                                         FinalizeOp finalize,
+                                                         ValueType init,
+                                                         ValueType *storage,
+                                                         KernelArgs... args)
+{
+    __shared__
+        UninitializedArray<ValueType, default_block_size / config::warp_size>
+            warp_partial;
+    static_assert(default_block_size / config::warp_size <= config::warp_size,
+                  "needs third reduction level");
+    auto tidx = thread::get_thread_id_flat<int64>();
+    auto grid_size = thread::get_thread_num_flat<int64>();
+    auto warp =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    auto partial = init;
+    for (int64 i = tidx; i < rows * cols; i += grid_size) {
+        const auto row = i / cols;
+        const auto col = i % cols;
+        partial = op(partial, fn(row, col, args...));
+    }
+    partial = reduce(warp, partial, op);
+    if (warp.thread_rank() == 0) {
+        warp_partial[threadIdx.x / config::warp_size] = partial;
+    }
+    __syncthreads();
+    if (threadIdx.x < config::warp_size) {
+        partial = reduce(warp,
+                         threadIdx.x < default_block_size / config::warp_size
+                             ? warp_partial[threadIdx.x]
+                             : init,
+                         op);
+        if (threadIdx.x == 0) {
+            storage[blockIdx.x] = finalize(partial);
+        }
+    }
+}
+
+
+template <typename ValueType, typename KernelFunction, typename ReductionOp,
+          typename FinalizeOp, typename... KernelArgs>
+void run_kernel_reduction(std::shared_ptr<const CudaExecutor> exec,
+                          KernelFunction fn, ReductionOp op,
+                          FinalizeOp finalize, ValueType init,
+                          ValueType *result, size_type size,
+                          KernelArgs &&... args)
+{
+    constexpr int oversubscription = 16;
+    gko::cuda::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = default_block_size;
+    const auto num_blocks = std::min<int64>(
+        ceildiv(size, block_size), exec->get_num_warps() * oversubscription);
```

Comment:
The difference doesn't really matter here, since the oversubscription is a tuning parameter. Since the reductions happen on three levels (warp, threadblock, grid), I thought it would make sense to use the same basis for handling the work distribution. This is probably only a difference of factor 2 anyways.